File: chunk.js
/* ============================================================
   chunk.js — page glue for chunk.html.

   Splits a Markdown document into retrieval-sized pieces along its own
   structure. Headings decide where sections are; size only decides where
   a section has to be cut. That ordering is the entire reason extract
   goes to the trouble of preserving headings — a chunk is a part of the
   document, not a 512-token window that happens to land somewhere.

   Every chunk carries heading_path, page and source, because that is
   what turns a retrieved fragment into a citation rather than an
   anonymous paragraph.

   Token counting here is toolkit.estimateTokens, a heuristic, because
   running BGE-M3's real tokenizer would mean a ~17MB download in a page
   that budgets 40KB for its own script. Every record says so in its
   `tokenizer` field, and the Python side re-checks. The heuristic errs
   high on purpose: an under-count makes a chunk that overflows the
   model's context, and an oversized chunk is not rejected at embed time
   — it is silently truncated, and its tail never reaches the vector.

   The Python twin is thehallucinatedlab/tools/chunk.py, and both run
   spec/chunk-fixtures.json. If they drift, both suites go red.

   Failure model:
     - Spec fails to load -> drop zone stays disabled, panel says why.
     - overlap >= max_tokens -> refused before any work, because each
       chunk would repeat the whole of the one before it.
     - Nothing to chunk -> says so rather than writing an empty file.
   ============================================================ */
(function () {
  'use strict';

  var toolkit = window.THL && window.THL.toolkit;

  /* @pure-start — no DOM, no network, no module state. */

  var FRONTMATTER = /^---\n([\s\S]*?)\n---\n?/;
  var PAGE_MARKER = /<!--\s*page:\s*(\d+)\s*-->/;
  var HEADING = /^(#{1,6})\s+(.*)$/;
  var FENCE = /^\s*```/;
  /* Deliberately simple. A full abbreviation list would be a lot of
     machinery for a boundary that only matters once a paragraph is
     already too big to keep whole. */
  var SENTENCE = /(?<=[.!?])\s+/;

  /* Only what extract writes: quoted strings and bare numbers, one per
     line. A general YAML parser would be a dependency for the sake of a
     header this project produced itself. */
  function parseFrontmatter(text) {
    var match = FRONTMATTER.exec(String(text || ''));
    if (!match) return { meta: {}, body: String(text || '') };

    var meta = {};
    var lines = match[1].split('\n');
    for (var i = 0; i < lines.length; i++) {
      var at = lines[i].indexOf(':');
      if (at < 0) continue;
      var key = lines[i].slice(0, at).trim();
      var value = lines[i].slice(at + 1).trim();
      if (value.length >= 2 && value.charAt(0) === '"' && value.charAt(value.length - 1) === '"') {
        meta[key] = value.slice(1, -1).replace(/\\"/g, '"').replace(/\\\\/g, '\\');
      } else if (/^-?\d+$/.test(value)) {
        meta[key] = parseInt(value, 10);
      } else if (value) {
        meta[key] = value;
      }
    }
    return { meta: meta, body: String(text).slice(match[0].length) };
  }

  /* Walks the document into blocks, each knowing the headings above it.

     Headings are not emitted as blocks: they become the path attached to
     everything beneath them. Emitting them too would put the heading in
     the output twice — once as the path, once as the first line of the
     chunk that follows it. */
  function parseBlocks(body) {
    var blocks = [];
    var stack = [];
    var page = null;
    var buffer = [];
    var fenced = false;

    function flush() {
      var text = buffer.join('\n').trim();
      buffer.length = 0;
      if (!text) return;
      var path = [];
      for (var i = 0; i < stack.length; i++) path.push(stack[i].title);
      blocks.push({ text: text, page: page, path: path });
    }

    var lines = String(body || '').split('\n');
    for (var i = 0; i < lines.length; i++) {
      var line = lines[i];

      if (FENCE.test(line)) {
        /* Inside a fence, blank lines and #-lines are code rather than
           structure; splitting on them would cut a snippet in half. */
        fenced = !fenced;
        buffer.push(line);
        continue;
      }
      if (fenced) { buffer.push(line); continue; }

      var marker = PAGE_MARKER.exec(line);
      if (marker && !line.replace(marker[0], '').trim()) {
        flush();
        page = parseInt(marker[1], 10);
        continue;
      }

      var heading = HEADING.exec(line);
      if (heading) {
        flush();
        var level = heading[1].length;
        while (stack.length && stack[stack.length - 1].level >= level) stack.pop();
        stack.push({ level: level, title: heading[2].trim() });
        continue;
      }

      if (!line.trim()) { flush(); continue; }
      buffer.push(line);
    }

    flush();
    return blocks;
  }

  function greedy(parts, joiner, budget, count) {
    var out = [];
    var current = '';
    for (var i = 0; i < parts.length; i++) {
      var candidate = current ? current + joiner + parts[i] : parts[i];
      if (current && count(candidate) > budget) {
        out.push(current);
        current = parts[i];
      } else {
        current = candidate;
      }
    }
    if (current) out.push(current);
    return out;
  }

  function nonEmpty(list) {
    var out = [];
    for (var i = 0; i < list.length; i++) if (String(list[i]).trim()) out.push(list[i]);
    return out;
  }

  /* Cuts one oversized block on the coarsest boundary that still fits,
     so a paragraph is only broken into sentences when the paragraph
     alone is too big, and into words only when one sentence is. */
  function splitText(text, budget, count) {
    if (count(text) <= budget) return [text];

    var attempts = [
      { parts: String(text).split('\n\n'), joiner: '\n\n' },
      { parts: String(text).split(SENTENCE), joiner: ' ' },
      { parts: String(text).split(' '), joiner: ' ' }
    ];

    for (var a = 0; a < attempts.length; a++) {
      var usable = nonEmpty(attempts[a].parts);
      if (usable.length < 2) continue;
      var out = greedy(usable, attempts[a].joiner, budget, count);
      var allFit = true;
      for (var i = 0; i < out.length; i++) {
        if (count(out[i]) > budget) { allFit = false; break; }
      }
      if (allFit) return out;
    }

    /* One enormous unbroken run. Cut on characters rather than emit
       something that would be truncated invisibly at embed time. */
    var approx = Math.max(1, budget * 3);
    var pieces = [];
    for (var c = 0; c < text.length; c += approx) pieces.push(text.slice(c, c + approx));
    return pieces;
  }

  /* The last `budget` tokens' worth of text, on a word boundary. */
  function tail(text, budget, count) {
    if (budget <= 0) return '';
    var words = String(text).split(/\s+/);
    var kept = [];
    for (var i = words.length - 1; i >= 0; i--) {
      kept.unshift(words[i]);
      if (count(kept.join(' ')) >= budget) break;
    }
    return kept.join(' ');
  }

  /* Compared element by element rather than by joining on a separator.
     Any separator is a character a heading could legitimately contain,
     so ["A B"] and ["A", "B"] would compare equal and two unrelated
     sections would silently merge into one chunk. */
  function samePath(a, b) {
    if (a.length !== b.length) return false;
    for (var i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
    return true;
  }

  function pack(blocks, options) {
    var count = options.count;
    var maxTokens = options.max_tokens;
    var chunks = [];

    /* Consecutive blocks sharing a heading path form a section, and a
       chunk never spans two. A piece that straddled a boundary could not
       name which heading it belonged to. */
    var sections = [];
    for (var b = 0; b < blocks.length; b++) {
      var last = sections[sections.length - 1];
      if (last && samePath(last.path, blocks[b].path)) {
        last.members.push(blocks[b]);
      } else {
        sections.push({ path: blocks[b].path.slice(), members: [blocks[b]] });
      }
    }

    function emit(prefix, path, body, page) {
      var trimmed = String(body || '').trim();
      if (!trimmed) return;
      var text = prefix + trimmed;
      chunks.push({
        text: text,
        source: options.source,
        heading_path: path.slice(),
        page: page === undefined ? null : page,
        chunk_index: chunks.length,
        token_count: count(text),
        tokenizer: options.tokenizer
      });
    }

    for (var s = 0; s < sections.length; s++) {
      var path = sections[s].path;
      var prefix = (options.heading_context && path.length) ? path.join(' > ') + '\n\n' : '';
      var budget = prefix ? maxTokens - count(prefix) : maxTokens;
      if (budget < 1) { prefix = ''; budget = maxTokens; }

      /* Flattened first, so every piece already fits and the loop below
         only has to decide where the boundaries fall. */
      var pieces = [];
      var members = sections[s].members;
      for (var m = 0; m < members.length; m++) {
        var parts = splitText(members[m].text, budget, count);
        for (var p = 0; p < parts.length; p++) {
          pieces.push({ text: parts[p], page: members[m].page });
        }
      }

      var current = '';
      var currentPage = null;

      for (var i = 0; i < pieces.length; i++) {
        var candidate = current ? current + '\n\n' + pieces[i].text : pieces[i].text;
        if (current && count(candidate) > budget) {
          emit(prefix, path, current, currentPage);
          var carry = tail(current, options.overlap, count);
          current = carry ? carry + '\n\n' + pieces[i].text : pieces[i].text;
          /* The carry was sized alone and the piece was checked alone;
             together they can exceed the budget. Drop the overlap rather
             than the guarantee. */
          if (count(current) > budget) current = pieces[i].text;
          currentPage = pieces[i].page;
        } else {
          current = candidate;
          if (currentPage === null) currentPage = pieces[i].page;
        }
      }
      emit(prefix, path, current, currentPage);
    }

    return chunks;
  }

  function toJsonl(chunks) {
    var lines = [];
    for (var i = 0; i < chunks.length; i++) lines.push(JSON.stringify(chunks[i]));
    return lines.length ? lines.join('\n') + '\n' : '';
  }

  /* The manifest can express each bound but not the relationship between
     two. An overlap at least as large as the chunk means every chunk
     begins with the whole of the previous one. */
  function overlapProblem(args) {
    if (args.overlap >= args.max_tokens) {
      return 'overlap (' + args.overlap + ') must be smaller than max_tokens (' +
             args.max_tokens + '); otherwise every chunk would repeat the whole of ' +
             'the one before it.';
    }
    return null;
  }

  function outputName(inputName) {
    var base = String(inputName || '').replace(/^.*[\\/]/, '');
    var dot = base.lastIndexOf('.');
    var stem = dot > 0 ? base.slice(0, dot) : base;
    stem = stem.replace(/[\x00-\x1f<>:"/\\|?*]/g, '').trim();
    if (!stem) stem = 'chunks';
    return stem + '.jsonl';
  }

  /* @pure-end */

  /* ---- DOM glue ---- */

  var manifest = null;
  var tool = null;
  var sourceFile = null;
  var downloadUrl = null;
  var busy = false;
  var el = {};

  function setStatus(message, kind) {
    if (!el.status) return;
    el.status.textContent = message || '';
    el.status.className = 'tool-status' + (kind ? ' is-' + kind : '');
  }

  function currentArgs() {
    return {
      max_tokens: parseInt(el.maxTokens.value, 10),
      overlap: parseInt(el.overlap.value, 10),
      tokenizer: 'estimate',
      heading_context: el.headingContext.checked
    };
  }

  function readAsText(file) {
    return new Promise(function (resolve, reject) {
      var reader = new FileReader();
      reader.onload = function () { resolve(String(reader.result)); };
      reader.onerror = function () { reject(new Error('That file could not be read.')); };
      reader.readAsText(file);
    });
  }

  function describe(chunks) {
    var counts = [];
    for (var i = 0; i < chunks.length; i++) counts.push(chunks[i].token_count);
    counts.sort(function (a, b) { return a - b; });
    var total = 0;
    for (var j = 0; j < counts.length; j++) total += counts[j];
    return {
      pieces: counts.length,
      total: total,
      median: counts.length ? counts[Math.floor(counts.length / 2)] : 0,
      largest: counts.length ? counts[counts.length - 1] : 0
    };
  }

  function offerDownload(jsonl, chunks, name) {
    if (downloadUrl) URL.revokeObjectURL(downloadUrl);
    downloadUrl = URL.createObjectURL(new Blob([jsonl], { type: 'application/jsonl' }));
    el.download.href = downloadUrl;
    el.download.download = name;
    el.resultName.textContent = name;

    var stats = describe(chunks);
    while (el.facts.firstChild) el.facts.removeChild(el.facts.firstChild);
    [['chunks', String(stats.pieces)],
     ['tokens', stats.total.toLocaleString() + ' (estimated)'],
     ['median', String(stats.median)],
     ['largest', String(stats.largest)]].forEach(function (pair) {
      var li = document.createElement('li');
      var label = document.createElement('span');
      label.textContent = pair[0];
      li.appendChild(label);
      li.appendChild(document.createTextNode(' ' + pair[1]));
      el.facts.appendChild(li);
    });

    /* First three records, pretty-printed. The file itself is one record
       per line; this is only so the shape is visible before download. */
    el.preview.textContent = chunks.slice(0, 3)
      .map(function (c) { return JSON.stringify(c, null, 2); })
      .join('\n\n') + (chunks.length > 3 ? '\n\n…' : '');
    el.result.hidden = false;
  }

  function run() {
    if (busy || !sourceFile) return;
    var validated = toolkit.validateArgs(currentArgs(), tool);
    if (!validated.ok) { setStatus(validated.errors.join(' '), 'error'); return; }

    var problem = overlapProblem(validated.args);
    if (problem) { setStatus(problem, 'error'); return; }

    busy = true;
    setStatus('Chunking…');

    readAsText(sourceFile).then(function (raw) {
      var text = String(raw).replace(/\r\n?/g, '\n');
      var parsed = parseFrontmatter(text);
      var blocks = parseBlocks(parsed.body);
      var chunks = pack(blocks, {
        max_tokens: validated.args.max_tokens,
        overlap: validated.args.overlap,
        heading_context: validated.args.heading_context,
        count: toolkit.estimateTokens,
        tokenizer: 'estimate',
        source: parsed.meta.source || sourceFile.name
      });

      if (!chunks.length) {
        setStatus('Nothing to chunk — the document had no body text.', 'error');
        return;
      }

      offerDownload(toJsonl(chunks), chunks, outputName(sourceFile.name));
      var headed = false;
      for (var i = 0; i < blocks.length; i++) if (blocks[i].path.length) { headed = true; break; }
      setStatus(headed
        ? 'Done — ' + chunks.length + ' chunks, sized by estimate. Re-check with thl for exact counts.'
        : 'Done — but no headings were found, so this fell back to size alone.', headed ? 'ok' : '');
    }).catch(function (err) {
      setStatus(err.message || 'Chunking failed.', 'error');
    }).then(function () {
      busy = false;
    });
  }

  function acceptFile(file) {
    if (!file) return;
    var max = (tool.input && tool.input.maxBytes) || Infinity;
    if (file.size > max) {
      setStatus('That file is ' + Math.round(file.size / 1048576) + 'MB; the limit is ' +
        Math.round(max / 1048576) + 'MB.', 'error');
      return;
    }
    sourceFile = file;
    el.dropTitle.textContent = file.name;
    el.dropHint.textContent = Math.max(1, Math.round(file.size / 1024)) + ' KB';
    el.run.disabled = false;
    setStatus('');
  }

  document.addEventListener('DOMContentLoaded', function () {
    el = {
      drop: document.getElementById('drop-zone'),
      dropTitle: document.getElementById('drop-title'),
      dropHint: document.getElementById('drop-hint'),
      pick: document.getElementById('pick-file'),
      file: document.getElementById('file-input'),
      maxTokens: document.getElementById('opt-max-tokens'),
      maxTokensValue: document.getElementById('max-tokens-value'),
      overlap: document.getElementById('opt-overlap'),
      overlapValue: document.getElementById('overlap-value'),
      headingContext: document.getElementById('opt-heading-context'),
      run: document.getElementById('chunk-btn'),
      status: document.getElementById('tool-status'),
      result: document.getElementById('tool-result'),
      resultName: document.getElementById('result-name'),
      preview: document.getElementById('result-preview'),
      facts: document.getElementById('result-facts'),
      download: document.getElementById('result-download'),
      args: document.getElementById('tool-args')
    };

    if (!toolkit) { setStatus('The tool runtime failed to load.', 'error'); return; }

    toolkit.loadManifest().then(function (loaded) {
      manifest = loaded;
      tool = toolkit.findTool(manifest, 'chunk');
      if (!tool) throw new Error('chunk is missing from the tool spec.');

      /* Ranges built from the manifest rather than the markup, so the
         page cannot offer a size the tool would reject. */
      tool.params.forEach(function (param) {
        var input = param.name === 'max_tokens' ? el.maxTokens
                  : param.name === 'overlap' ? el.overlap : null;
        if (!input) return;
        input.min = param.min;
        input.max = param.max;
        input.value = param.default;
      });
      el.maxTokensValue.textContent = el.maxTokens.value;
      el.overlapValue.textContent = el.overlap.value;

      if (el.args) toolkit.renderParamTable(tool, el.args);
      el.drop.classList.remove('is-disabled');
      setStatus('');
    }).catch(function (err) {
      setStatus(err.message || 'The tool spec failed to load.', 'error');
    });

    el.pick.addEventListener('click', function () { el.file.click(); });
    el.file.addEventListener('change', function () { acceptFile(el.file.files[0]); });
    el.run.addEventListener('click', run);
    el.maxTokens.addEventListener('input', function () {
      el.maxTokensValue.textContent = el.maxTokens.value;
    });
    el.overlap.addEventListener('input', function () {
      el.overlapValue.textContent = el.overlap.value;
    });

    ['dragenter', 'dragover'].forEach(function (name) {
      el.drop.addEventListener(name, function (event) {
        event.preventDefault();
        el.drop.classList.add('is-over');
      });
    });
    ['dragleave', 'drop'].forEach(function (name) {
      el.drop.addEventListener(name, function (event) {
        event.preventDefault();
        el.drop.classList.remove('is-over');
      });
    });
    el.drop.addEventListener('drop', function (event) {
      acceptFile(event.dataTransfer.files[0]);
    });
  });

})();

File: python/thehallucinatedlab/tools/chunk.py
"""Split a Markdown document into retrieval-sized pieces.

Structure first, size second. A chunk is a section of the document --
identified by the headings above it and the page it started on -- rather
than an arbitrary window of characters that happens to be 512 tokens
long. That distinction is the whole reason ``extract`` goes to the
trouble of preserving headings.

Size is the constraint, not the organising principle. Within a section,
blocks are packed until the next one would not fit; only a single block
that is itself too large gets cut, and then on the largest boundary that
works: paragraphs, then sentences, then words.

Every chunk records where it came from -- ``heading_path``, ``page``,
``source`` -- because that metadata is what turns a retrieved fragment
into a citation. Output is JSONL, one record per line, which is the same
shape a LangChain ``Document`` carries and stays greppable besides.

One honest limitation. The browser cannot run BGE-M3's tokenizer without
a 17MB download, so it chunks against a heuristic that deliberately
over-counts and records ``tokenizer: "estimate"`` on every record.
``embed`` re-checks rather than trusting it. Under-counting would be the
expensive direction: an oversized chunk is not rejected at embed time,
it is silently truncated, and the tail never reaches the vector.
"""

from __future__ import annotations

import json
import os
import re
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Union

from ..errors import InvalidArgument
from ..registry import registry
from .tokenize import counter_for, is_exact

Source = Union[str, "os.PathLike[str]", bytes, bytearray, Any]

_FRONTMATTER = re.compile(r"\A---\n(.*?)\n---\n?", re.DOTALL)
_PAGE_MARKER = re.compile(r"<!--\s*page:\s*(\d+)\s*-->")
_HEADING = re.compile(r"^(#{1,6})\s+(.*)$")
_FENCE = re.compile(r"^\s*```")
# Split after . ! ? that are followed by whitespace. Deliberately simple:
# a full abbreviation list would be a lot of machinery for a boundary
# that only matters when one paragraph is already oversized.
_SENTENCE = re.compile(r"(?<=[.!?])\s+")


@dataclass
class Chunk:
    """One piece, and everything needed to cite it."""

    text: str
    source: str
    heading_path: list[str]
    page: int | None
    chunk_index: int
    token_count: int
    tokenizer: str

    def as_dict(self) -> dict[str, Any]:
        return {
            "text": self.text,
            "source": self.source,
            "heading_path": self.heading_path,
            "page": self.page,
            "chunk_index": self.chunk_index,
            "token_count": self.token_count,
            "tokenizer": self.tokenizer,
        }


@dataclass
class ChunkResult:
    """What a chunking produced."""

    chunks: list[Chunk]
    tokenizer: str
    exact: bool
    max_tokens: int
    path: Path | None = None
    warnings: list[str] = field(default_factory=list)

    @property
    def total_tokens(self) -> int:
        return sum(c.token_count for c in self.chunks)

    def as_jsonl(self) -> str:
        # ensure_ascii=False so non-Latin text stays readable in the file
        # rather than becoming \uXXXX soup nobody can grep.
        return "".join(
            json.dumps(c.as_dict(), ensure_ascii=False) + "\n" for c in self.chunks
        )

    def __str__(self) -> str:
        if self.exact:
            how = "exact"
        elif self.tokenizer == "estimate":
            how = "estimated"
        else:
            how = f"{self.tokenizer}, approximate for bge"
        where = str(self.path) if self.path else "in memory"
        return (
            f"{len(self.chunks)} chunk(s), {self.total_tokens:,} tokens "
            f"[{how}] -> {where}"
        )


def parse_frontmatter(text: str) -> tuple[dict[str, Any], str]:
    """Split extract's YAML block off the front.

    Only handles what ``extract`` writes -- quoted strings and bare
    numbers, one per line. A general YAML parser would be a dependency
    for the sake of a header this package produced itself.
    """
    match = _FRONTMATTER.match(text)
    if not match:
        return {}, text

    meta: dict[str, Any] = {}
    for line in match.group(1).split("\n"):
        if ":" not in line:
            continue
        key, _, raw = line.partition(":")
        value = raw.strip()
        if len(value) >= 2 and value[0] == '"' and value[-1] == '"':
            meta[key.strip()] = value[1:-1].replace('\\"', '"').replace("\\\\", "\\")
        elif re.fullmatch(r"-?\d+", value):
            meta[key.strip()] = int(value)
        elif value:
            meta[key.strip()] = value
    return meta, text[match.end():]


def parse_blocks(body: str) -> list[dict[str, Any]]:
    """Walk the document into blocks, each knowing its heading path.

    Headings are not emitted as blocks. They become the path attached to
    everything beneath them, so the text is not duplicated -- once as a
    heading and again inside the chunk that follows it.
    """
    blocks: list[dict[str, Any]] = []
    stack: list[tuple[int, str]] = []
    page: int | None = None
    buffer: list[str] = []
    fenced = False

    def flush() -> None:
        text = "\n".join(buffer).strip()
        buffer.clear()
        if text:
            blocks.append({
                "text": text,
                "page": page,
                "path": [title for _, title in stack],
            })

    for line in body.split("\n"):
        if _FENCE.match(line):
            # Inside a fence, blank lines and #-lines are code, not
            # structure. Splitting there would cut a snippet in half.
            fenced = not fenced
            buffer.append(line)
            continue

        if fenced:
            buffer.append(line)
            continue

        marker = _PAGE_MARKER.search(line)
        if marker and not line.strip().replace(marker.group(0), "").strip():
            flush()
            page = int(marker.group(1))
            continue

        heading = _HEADING.match(line)
        if heading:
            flush()
            level = len(heading.group(1))
            while stack and stack[-1][0] >= level:
                stack.pop()
            stack.append((level, heading.group(2).strip()))
            continue

        if not line.strip():
            flush()
            continue

        buffer.append(line)

    flush()
    return blocks


def _greedy(
    parts: list[str], joiner: str, budget: int, count: Callable[[str], int]
) -> list[str]:
    """Fill each piece until the next part would not fit."""
    out: list[str] = []
    current = ""
    for part in parts:
        candidate = f"{current}{joiner}{part}" if current else part
        if current and count(candidate) > budget:
            out.append(current)
            current = part
        else:
            current = candidate
    if current:
        out.append(current)
    return out


def _split_text(text: str, budget: int, count: Callable[[str], int]) -> list[str]:
    """Cut one oversized block on the largest boundary that still fits.

    Tried in order, coarsest first, so a paragraph is only broken into
    sentences when the paragraph alone is too big, and only into words
    when a single sentence is.
    """
    if count(text) <= budget:
        return [text]

    for parts, joiner in (
        (text.split("\n\n"), "\n\n"),
        (_SENTENCE.split(text), " "),
        (text.split(" "), " "),
    ):
        usable = [p for p in parts if p.strip()]
        if len(usable) < 2:
            continue
        out = _greedy(usable, joiner, budget, count)
        if all(count(piece) <= budget for piece in out):
            return out

    # Nothing divides small enough -- one enormous unbroken run. Cut on
    # characters rather than emit something that would be truncated
    # invisibly at embed time, taking its tail out of the vector.
    approx = max(1, budget * 3)
    return [text[i:i + approx] for i in range(0, len(text), approx)]


def _tail(text: str, budget: int, count: Callable[[str], int]) -> str:
    """The last ``budget`` tokens' worth of ``text``, on a word boundary."""
    if budget <= 0:
        return ""
    words = text.split()
    kept: list[str] = []
    for word in reversed(words):
        kept.insert(0, word)
        if count(" ".join(kept)) >= budget:
            break
    return " ".join(kept)


def pack(
    blocks: list[dict[str, Any]],
    *,
    max_tokens: int,
    overlap: int,
    heading_context: bool,
    count: Callable[[str], int],
    source: str,
) -> list[Chunk]:
    """Group blocks into chunks that respect both structure and size."""
    chunks: list[Chunk] = []

    def prefix_for(path: list[str]) -> str:
        return " > ".join(path) + "\n\n" if (heading_context and path) else ""

    # Consecutive blocks sharing a heading path form a section. A chunk
    # never spans two sections, so a piece is always answerable by one
    # heading path rather than an ambiguous pair.
    sections: list[tuple[list[str], list[dict[str, Any]]]] = []
    for block in blocks:
        if sections and sections[-1][0] == block["path"]:
            sections[-1][1].append(block)
        else:
            sections.append((list(block["path"]), [block]))

    def emit(prefix: str, path: list[str], body: str, page: int | None) -> None:
        body = body.strip()
        if not body:
            return
        text = prefix + body
        chunks.append(Chunk(
            text=text,
            source=source,
            heading_path=list(path),
            page=page,
            chunk_index=len(chunks),
            token_count=count(text),
            tokenizer="",     # filled in by the caller, which knows it
        ))

    for path, members in sections:
        prefix = prefix_for(path)
        budget = max_tokens - count(prefix) if prefix else max_tokens
        if budget < 1:
            # A heading path so long it leaves no room for text. Drop the
            # prefix for this section rather than emit empty chunks.
            prefix, budget = "", max_tokens

        # Flatten first: every piece here already fits the budget, so the
        # packing loop below only has to decide where boundaries fall.
        pieces = [
            (piece, block["page"])
            for block in members
            for piece in _split_text(block["text"], budget, count)
        ]

        current = ""
        current_page: int | None = None

        for piece, page in pieces:
            candidate = f"{current}\n\n{piece}" if current else piece
            if current and count(candidate) > budget:
                emit(prefix, path, current, current_page)
                carry = _tail(current, overlap, count)
                current = f"{carry}\n\n{piece}" if carry else piece
                # The carry is sized in tokens, and the piece was only
                # checked on its own. Together they can exceed the
                # budget, which would hand embed an oversized chunk --
                # so drop the overlap rather than the guarantee.
                if count(current) > budget:
                    current = piece
                current_page = page
            else:
                current = candidate
                if current_page is None:
                    current_page = page

        emit(prefix, path, current, current_page)

    return chunks


def chunk(
    source: Source,
    dest: str | os.PathLike[str] | None = None,
    *,
    max_tokens: int | None = None,
    overlap: int | None = None,
    tokenizer: str | None = None,
    heading_context: bool | None = None,
    filename: str | None = None,
) -> ChunkResult:
    """Split a Markdown document into JSONL chunks.

    Args:
        source: Path to a Markdown file, raw bytes, a binary file object,
            or the document as a string.
        dest: Where to write the JSONL. Defaults to the source path with
            a .jsonl extension.
        max_tokens: Largest chunk, in tokens. Default 512.
        overlap: Tokens repeated across a boundary. Default 64. Must be
            smaller than ``max_tokens``.
        tokenizer: ``"bge"`` (default), ``"openai"`` or ``"estimate"``.
        heading_context: Prepend the heading path to each chunk.
        filename: Name to attribute bytes to.

    Returns:
        A :class:`ChunkResult`.

    Raises:
        InvalidArgument: an argument is outside what the manifest allows,
            or overlap is not smaller than max_tokens.
        DependencyMissing: the chosen tokenizer needs the ``chunk`` extra.
    """
    args = registry.validate(
        "chunk",
        max_tokens=max_tokens,
        overlap=overlap,
        tokenizer=tokenizer,
        heading_context=heading_context,
    )

    # The manifest can express each bound but not the relationship. An
    # overlap at least as large as the chunk means every chunk begins
    # with the whole of the one before it, and the packer never advances.
    if args["overlap"] >= args["max_tokens"]:
        raise InvalidArgument(
            f"overlap ({args['overlap']}) must be smaller than max_tokens "
            f"({args['max_tokens']}); otherwise each chunk would repeat the whole "
            "of the previous one and chunking would not terminate."
        )

    source_path: Path | None = None
    if isinstance(source, str) and "\n" in source:
        text = source                      # the document itself
    elif isinstance(source, bytes | bytearray):
        text = bytes(source).decode("utf-8", errors="replace")
    elif hasattr(source, "read"):
        raw = source.read()
        if not isinstance(raw, bytes):
            raise InvalidArgument("A file object must be opened in binary mode.")
        text = raw.decode("utf-8", errors="replace")
    else:
        source_path = Path(os.fspath(source))
        if not source_path.is_file():
            raise InvalidArgument(f"No such document: {source_path}")
        text = source_path.read_bytes().decode("utf-8", errors="replace")

    text = re.sub(r"\r\n?", "\n", text)
    meta, body = parse_frontmatter(text)
    name = (
        filename
        or (str(meta.get("source")) if meta.get("source") else None)
        or (source_path.name if source_path else "")
    )

    count = counter_for(args["tokenizer"])
    blocks = parse_blocks(body)
    chunks = pack(
        blocks,
        max_tokens=args["max_tokens"],
        overlap=args["overlap"],
        heading_context=args["heading_context"],
        count=count,
        source=name,
    )
    for piece in chunks:
        piece.tokenizer = args["tokenizer"]

    warnings: list[str] = []
    if not is_exact(args["tokenizer"]):
        warnings.append(
            f"Counted with {args['tokenizer']}, which does not match the embedding "
            "model. Sizes are approximate; embed will re-check."
        )
    if not blocks:
        warnings.append("Nothing to chunk -- the document had no body text.")
    if blocks and not any(b["path"] for b in blocks):
        warnings.append(
            "No headings found, so chunking fell back to size alone. If this came "
            "from a PDF without an outline, that is expected."
        )

    target: Path | None = None
    if dest is not None:
        target = Path(os.fspath(dest))
    elif source_path is not None:
        target = source_path.with_suffix(".jsonl")

    result = ChunkResult(
        chunks=chunks,
        tokenizer=args["tokenizer"],
        exact=is_exact(args["tokenizer"]),
        max_tokens=args["max_tokens"],
        path=target,
        warnings=warnings,
    )

    if target is not None:
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(result.as_jsonl(), encoding="utf-8", newline="")

    return result


registry.register("chunk", chunk)
